File: RegresionLineal.py
### Regresion Lineal ###

#Librerias

import numpy as np

from sklearn.linear_model import LinearRegression

from matplotlib import pyplot as plt

import statsmodels.api as sm

import pandas as pd

from statsmodels.formula.api import ols

from statsmodels.stats.stattools import durbin_watson

import statsmodels.stats.api as sms

from statsmodels.compat import lzip

from statsmodels.stats.diagnostic import het_white

from patsy import dmatrices

from statsmodels.stats.outliers_influence import variance_inflation_factor

from chow_test import chowtest

#Funciones

#Residual Squared Sum (RSE)

def rss(mod):
    print('Residual Squared Sum (RSE)',sep = '\n')
    return(np.sum(mod.resid**2))

#Residual Standard Error (RSS)

def rse(mod):
    scr = np.sum(mod.resid**2)
    n = len(mod.resid)
    print('Residual Standard Error (RSS)',sep = '\n')
    return(np.sqrt(scr/(n-2)))

#Mean Absolute Error (MAE)

def mae(mod):
    print('Mean Absolute Error (MAE)',sep = '\n')
    return(np.sum(abs(mod.resid))/(len(mod.resid)))

#Mean Absolute Percent Error (MAPE)

def mape(mod,df,y):
    sal = np.round(np.mean(np.abs(100*mod.resid/df.y)), 2)
    print('Mean Absolute Percent Error (MAPE)',sep = '\n')
    return(sal)

#Mahalanobis distance

def mahalanobis(x=None, data=None, cov=None):

    x_mu = x - np.mean(data)
    if not cov:
        cov = np.cov(data.values.T)
    inv_covmat = np.linalg.inv(cov)
    left = np.dot(x_mu, inv_covmat)
    mahal = np.dot(left, x_mu.T)
    return mahal.diagonal()

### Regresion Lineal Simple ###

#Datos de ejemplo

x = np.array([5, 15, 25, 35, 45, 55]).reshape((-1, 1)) #una columna y n filas

y = np.array([5, 20, 14, 32, 22, 38])

#Nombramos funcion

model = LinearRegression() #Nombramos la funcion LinearRegression como model

#Estimar modelo

model.fit(x, y)

model = LinearRegression().fit(x, y) #mas corto

#Bondad del ajuste R^2

r_sq = model.score(x, y)

print('coefficient of determination:', r_sq)

#Parametros

#Posicion
print('intercept:', model.intercept_)

#Pendiente
print('slope:', model.coef_)

#Estimaciones

y_pred = model.predict(x)

print('predicted response:', y_pred, sep='\n')

#Manualmente
y_pred = model.intercept_ + model.coef_ * x

print('predicted response:', y_pred, sep='\n')

#Preciccion (con datos nuevos)

x_new = np.arange(5).reshape((-1, 1))

print(x_new)

y_new = model.predict(x_new) #Predicciones con los nuevos valores de x

print(y_new) 

#Grafico

plt.plot(x,y_pred,color='red') #Recta de regresion

plt.scatter(x,y,color='black') #Grafico de dispersion

plt.show()

### Regresion Lineal Multiple ###

# Datos de ejemplo

x = [[0, 1], [5, 1], [15, 2], [25, 5], [35, 11], [45, 15], [55, 34], [60, 35]] #dos variables predictoras

y = [4, 5, 20, 14, 32, 22, 38, 43]

x, y = np.array(x), np.array(y)

# Estimacion

model = LinearRegression().fit(x, y)

#Bondad del ajuste R^2

r_sq = model.score(x, y)

print('coefficient of determination:', r_sq)

#Posicion y pendientes

print('intercept:', model.intercept_)

print('slope:', model.coef_)

#Estimaciones

y_pred = model.predict(x)

print('predicted response:', y_pred, sep='\n')

#Prediccion con nuevos datos

x_new = np.arange(10).reshape((-1, 2))

print(x_new)

y_new = model.predict(x_new)

print(y_new)

#Con libreria statmodels

#Trabajar con tablas de datos

#Tabla de datos con pandas

x = np.array([0,5,15,25,35,45,55,60])

v = np.array([1,1,2,5,11,15,34,35])

y = np.array([4, 5, 20, 14, 32, 22, 38, 43])

df = pd.DataFrame({'x1': x,'x2':v, 'y':y})

#Agregamos variable de unos #Para tener posicion

df = sm.add_constant(df)

#Estimacion

model = ols('y ~ x1 + x2', data=df).fit()

#Resumen del modelo

print(model.summary()) #Estadisticos del modelo

#Estimaciones

estimaciones = model.fittedvalues

print('predicted response:', model.fittedvalues, sep='\n')

print('predicted response:', model.predict(df), sep='\n')

#Residuos 

residuos = model.resid

print('predicted response:', model.resid, sep='\n')

#Prediccion con valores nuevos

x2 = np.array([75,85,100])

v2 = np.array([37,38,40])

x_new = pd.DataFrame({'x1': x2,'x2':v2})

x_new = sm.add_constant(x_new)

y_new = model.predict(x_new)

print(y_new)

#Intervalo de confianza

print ('Confidence intervals',model.conf_int(0.01))   
# 99% confidence interval

## Validacion del modelo

dir(model) #todos los estadisticos

#Bondad del ajuste R^2

print('coefficient of determination:', model.rsquared)

print('adjusted coefficient of determination:', model.rsquared_adj)

#Parametros

print('regression coefficients:', model.params)

#Grafico de residuos

#Residuos vs predictor

#Envergadura del grafico
fig = plt.figure(figsize=(12,8))

#Grafico
fig = sm.graphics.plot_regress_exog(model, 'x1', fig=fig)

#Residuos vs y

plt.scatter(df.y,residuos) #Grafico de dispersion
plt.xlabel('Y')
plt.axhline(y=0, color='r', linestyle='-')

plt.show()

#Residual Squared Sum (RSS)

np.sum(model.resid**2)

rss(model) #con funcion propia

#Residual Standard Error (RSE) 

np.sqrt(np.sum(model.resid**2)/(len(model.resid)-2))

rse(model)

#Mean Squared Error (MSE)

model.mse_total

#Anova

table=sm.stats.anova_lm(model)

print(table)

#Mean Absolute Error (MAE)

np.sum(abs(model.resid))/(len(model.resid))

mae(model)

#Mean Absolute Percent Error (MAPE)

np.round(np.mean(np.abs(100*model.resid/df.y)), 2)

mape(model,df,y)

## No linealidad ##

#Residuos vs y

plt.scatter(df.y,residuos) #Grafico de dispersion
plt.xlabel('Y')
plt.axhline(y=0, color='r', linestyle='-')

plt.show()

#Q-Q plot

sm.qqplot(model.resid,fit=True,line='45')
plt.show()

## Autocorrelacion ##

#Contraste de Durbin Watson

durbin_watson(model.resid)

## Heterocedasticidad ##

#Goldfelt Quant test

sms.het_goldfeldquandt(model.resid, model.model.exog)

#Test de White

white_test = het_white(model.resid,  model.model.exog)

#define labels to use for output of White's test
labels = ['Test Statistic', 'Test Statistic p-value',
          'F-Statistic', 'F-Test p-value']

#print results of White's test
print(dict(zip(labels, white_test)))

#Breusch-Pagan test

names = ['Lagrange multiplier statistic', 'p-value',
        'f-value', 'f p-value']
test = sms.het_breuschpagan(model.resid, model.model.exog)

lzip(names, test)

## Valores atipicos ##

#High leverage h

fig, ax = plt.subplots(figsize=(12,8))
fig = sm.graphics.influence_plot(model, alpha  = 0.05, 
                                 ax = ax, criterion="cooks")

#Distancia de Cook

np.set_printoptions(suppress=True)

#create instance of influence
influence = model.get_influence()

#obtain Cook's distance for each observation
cooks = influence.cooks_distance

#display Cook's distances
print(cooks)

#Grafico

plt.scatter(cooks[0], df.y)
plt.ylabel('y')
plt.xlabel('Cooks Distance')
plt.show()

#Distancia de Mahalanobis

#Necesitamos datos sin constante

x = np.array([0,5,15,25,35,45,55,60])
v = np.array([1,1,2,5,11,15,34,35])
y = np.array([4, 5, 20, 14, 32, 22, 38, 43])
dfm = pd.DataFrame({'x1': x,'x2':v, 'y':y})

#Agregar distancia a la base de datos
mha = mahalanobis(x=dfm, data=dfm[['x1', 'x2','y']])

print(mha)

#Grafico

plt.scatter(mha, range(0,len(df.y)))
plt.ylabel('Observation')
plt.xlabel('Mahalanobis Distance')
plt.show()

#range(0,len(df.y)) indicador de observacion

## Multicolinealidad ##

#matrix for linear regression model using 'y' as response variable 
Y, X = dmatrices('y ~ x1 + x2', data=df, return_type='dataframe')

#calculate VIF for each explanatory variable
vif = pd.DataFrame()
vif['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
vif['variable'] = X.columns

#VIF for each explanatory variable 
vif

## Cambio estructural ##

#Datos con cambio estructural

dfe = pd.DataFrame({'x': [1, 1, 2, 3, 4, 4, 5, 5, 6, 7, 7, 8, 8, 9, 10, 10,
                         11, 12, 12, 13, 14, 15, 15, 16, 17, 18, 18, 19, 20, 20],
                   'y': [3, 5, 6, 10, 13, 15, 17, 14, 20, 23, 25, 27, 30, 30, 31,
                         33, 32, 32, 30, 32, 34, 34, 37, 35, 34, 36, 34, 37, 38, 36]})

plt.plot(dfe.x, dfe.y, 'o')

#Chow Breakpoint Test

chowtest(y=dfe[['y']], X=dfe[['x']],
         last_index_in_model_1=15,
         first_index_in_model_2=16,
         significance_level=.05)

#CUMSUM

File: Regresion/RegresionLineal.R
#### Regresion lineal ####

#Librerias

library(MASS)
library(ISLR2)
library(corrplot)
library(tidyverse)
library(car)
library(readr)
library(stargazer)
library(Metrics)
library(lmtest)
library(car)
library(wooldridge)
library(qcc)
library(strucchange)

#Funciones

#Mean Squared Error MSE
mse <- function(mod) 
  mean(mod$residuals^2)

#Rango hat
#Da el rango para comparar el estadistico h
rango.hat = function(datos){
  return(2.5*((dim(datos)[2]+1)/dim(datos)[1]))
}

#Funcion para evaluar p-valor

tStest = function(x){
  
  nom = names(x)
  a = rep(0,length(x))
  
  for(i in seq_along(x)){
    
    if(x[i] > 0.025){a[i]='Se acepta Ho: Variable no significativa'}
    
    else{a[i]='Se rechaza Ho: Variable es significativa'}
    
  }
  
  return(tibble('Variables'=nom,'Contraste'=a,"p-valor"=x))
  
}

#Funcion para obtener como texto la ecuacion del modelo

model_equation <- function(model, ...) {
  format_args <- list(...)
  
  model_coeff <- model$coefficients
  format_args$x <- abs(model$coefficients)
  model_coeff_sign <- sign(model_coeff)
  model_coeff_prefix <- case_when(model_coeff_sign == -1 ~ " - ",
                                  model_coeff_sign == 1 ~ " + ",
                                  model_coeff_sign == 0 ~ " + ")
  model_eqn <- paste(strsplit(as.character(model$call$formula), "~")[[2]], # 'y'
                     "=",
                     paste(if_else(model_coeff[1]<0, "- ", ""),
                           do.call(format, format_args)[1],
                           paste(model_coeff_prefix[-1],
                                 do.call(format, format_args)[-1],
                                 " * ",
                                 names(model_coeff[-1]),
                                 sep = "", collapse = ""),
                           sep = ""))
  return(model_eqn)
}

#### Regresion Lineal Simple ####

#Obtencion de datos y analisis explotatorio

datos = Boston #Obtenemos los datos desde ISLR2

?Boston #Informacion sobre los datos

head(Boston) #Primeras filas

summary(Boston) #Estadistica descriptiva

pairs(Boston) #Graficos de dispersion multiples

(correlaciones = cor(Boston)) #Matriz de correlaciones

corrplot(correlaciones,method='ellipse') #Grafico de las correlaciones

cov(Boston) #Matriz de covarianzas

#Estimacion del modelo

lm.fit <- lm(medv ~ lstat , data = datos)

summary(lm.fit) #Resumen #t #F #R^2 #Estadistica descriptiva de error

names(lm.fit) #Vemos que objetos contiene

lm.fit$coefficients #coeficientes

lm.fit$residuals #residuos

lm.fit$fitted.values #estimaciones

lm.fit$rank #rango

lm.fit$model #explicada y explicativa

coef(lm.fit) #coeficientes

model_equation(lm.fit) #Ecuacion del modelo

#Agregamos las estimaciones a los datos 

datos = tibble(datos,'estimaciones'=lm.fit$fitted.values)

#Intervalo de confianza (para los coeficientes)

confint(lm.fit)

#Predicciones
#Con intervalos de confianza y de prediccion

#Prediccion para valores 5, 10 y 15
predict(lm.fit , data.frame(lstat = (c(5, 10, 15))))

#Prediccion e intervalo de confianza
#Valores entre los que se encuentra la funcion lineal real a un 95%
predict(lm.fit , data.frame(lstat = (c(5, 10, 15))),
        interval = "confidence")

#Prediccion e intervalo de estimacion
#Valores entre los que se encuentra el valor real de y a un 95%
predict(lm.fit , data.frame(lstat = (c(5, 10, 15))),
        interval = "prediction")

#Grafico

plot(datos$lstat , datos$medv) #grafico de dispersion
abline(lm.fit) #recta de regresion

#Grafico personalizado
plot(datos$lstat , datos$medv, xlab='X', ylab='Y', main='Regresion') 
abline(lm.fit, col='red') #recta de regresion
#Leyenda
legend(legend=c("Observaciones","Regresion"),
       fill=c("black", "red"), x = "bottomright")

#abline() para pendiente y posicion arbitraria

abline(a = 20, b=1 )

#Grafico ggplot()

#Grafico de dispersi?n para dos variables
ggplot(data = datos) +
  geom_point(mapping = aes(x = lstat, y = medv)) +
  geom_line(mapping = aes(x=lstat, y = estimaciones), colour = 'red',lwd=0.75) +
  geom_smooth(mapping = aes(x=lstat, y = medv), 
              colour = 'blue',lwd=0.5, se=FALSE)+
  ggtitle('Regresion Lineal')+
  labs(x='X',y='Y')

#Grafico de regresion lineal con intervalos de confianza

ggplot(datos, aes(x=lstat, y=medv)) + 
  geom_point(color='black') + 
  geom_smooth(method=lm, color='red')

#Grafico con intervalos de prediccion

intv = cbind(Boston, predict(lm.fit, interval = "prediction"))

ggplot(intv, aes(x = lstat,y=medv)) +
  geom_ribbon(aes(ymin = lwr, ymax = upr), fill = "gray", alpha = 0.5) +
  geom_point(color='black') +
  geom_line(aes(y = fit), colour = "blue", size = 1)

#En azul la funcion real

#Separamos pantalla 

par(mfrow = c(2, 2))

#Graficamos los residuos

plot(lm.fit)

#Residuos y residuos stunderizados

plot(predict(lm.fit), residuals(lm.fit))
plot(predict(lm.fit), rstudent(lm.fit))
#Hay evidencia de no linearidad

#Estadistico h (High-Leverage)

hatvalues(lm.fit)

plot(hatvalues(lm.fit))

which.max(hatvalues(lm.fit))

#Validacion del modelo

#R^2 manualmente
RSS <- sum(summary(lm.fit)$residuals^2)
TSS <- sum((datos$medv - mean(datos$medv))^2)
(R2 <- 1 - RSS/TSS)

#R^2 sacado de lm()

summary(lm.fit)$r.sq

#RSE Residual Standard Error

#RSE manualmente

n <- nrow(datos)
(RSE <- sqrt(RSS / (n-2)))

#RSE sacado de lm()

summary(lm.fit)$sigma

#R^2 ajustado

summary(lm.fit)$adj.r.squared

#Test F

(testF = summary(lm.fit)$fstatistic)

pf(testF[1],testF[2],testF[3],lower.tail = FALSE) #p-valor

#Parametros y test t

summary(lm.fit)$coefficients

names(summary(lm.fit)) #objetos contenidos en summary

summary(lm.fit)[["coefficients"]][, "t value"] #valor test

summary(lm.fit)$coefficients[,3] #valor test

summary(lm.fit)[["coefficients"]][, "Pr(>|t|)"] #p-valor

summary(lm.fit)$coefficients[,4] #p-valor

tStest(summary(lm.fit)$coefficients[,4])

#codigo para extraer objetos del summary

#MSE Mean Squared Error

#Manual
(MSE = sum((datos$medv - datos$estimaciones)^2)/nrow(datos))

#Con funcion

mse(lm.fit)

#Con ANOVA

anova(lm.fit)['Residuals', 'Mean Sq']

#### Regresion Lineal Multiple ####

#Estimacion con varias variables

lm.fit <- lm(medv ~ lstat + age , data = Boston)
#Si queremos usar todas usamos: y ~ .

summary(lm.fit)

#Si queremos usar todas usamos: y ~ .

lm.fit <- lm(medv~., data = Boston)

summary(lm.fit)

#R^2 

summary(lm.fit)$r.sq

#RSE

summary(lm.fit)$sigma

#F

summary(lm.fit)$f

#Multicolinealidad

vif(lm.fit)

#Quitar una variable

lm.fit1 <- lm(medv ~ . - age, data = Boston)
summary(lm.fit1)

#Alternativamente

lm.fit1 <- update(lm.fit , ~ . - age)

#Interaccion entre variables

summary(lm(medv ~ lstat * age , data = Boston))

#Todas las variables mas una interaccion
summary(lm(medv ~. + lstat * age , data = Boston))

#Transformaciones no lineales

lm.fit2 <- lm(medv ~ lstat + I(lstat^2), data = Boston) #I(x^k) para predictor^k
summary(lm.fit2)

#Anova para comparar que modelo es mejor

lm.fit <- lm(medv ~ lstat, data = Boston)
anova(lm.fit , lm.fit2)
#Contrastamos que hay diferencia de error
#El modelo no lineal es superior #coherente a la evidencia de no linealidad

#Grafico de error

par(mfrow = c(2, 2))
plot(lm.fit2)
#Continua habiendo patron en el error
#El cambio en la pendiente de la Normal Q-Q indica variable sin identificar

#Regresion polinomica de grado 5
#lstat + lstat^2 + ... lstat^5 como predictores

lm.fit5 <- lm(medv ~ poly(lstat , 5), data=Boston)
summary(lm.fit5)

#Alternativamente

summary(lm(medv ~ lstat+I(lstat^2)+I(lstat^3)+I(lstat^4)+I(lstat^5),data=Boston))

#Predictores cualitativos

#Datos

head(Carseats)

?Carseats

#Estimacion

lm.fit <- lm(Sales ~ . + Income:Advertising + Price:Age,
             data = Carseats)
summary(lm.fit)

#Composicion de la dummy
contrasts(Carseats$ShelveLoc)

#Precision del modelo

#Validacion del modelo

#R^2 manualmente
RSS <- sum(summary(lm.fit)$residuals^2)
TSS <- sum((datos$medv - mean(datos$medv))^2)
(R2 <- 1 - RSS/TSS)

#R^2 sacado de lm()

summary(lm.fit)$r.sq

#RSE Residual Standard Error

#RSE manualmente

n <- nrow(datos)
(RSE <- sqrt(RSS / (n-2)))

#RSE sacado de lm()

summary(lm.fit)$sigma

#R^2 ajustado

summary(lm.fit)$adj.r.squared

#Test F

summary(lm.fit)$fstatistic

#Parametros y test t

summary(lm.fit)$coefficients

#MSE Mean Squared Error

#Manual
(MSE = sum((datos$medv - datos$estimaciones)^2)/nrow(datos))

#Con ANOVA

anova(lm.fit)['Residuals', 'Mean Sq']

#Con funcion

mse(lm.fit)

#Error Medio Absoluto (EMA)

mae(Carseats$Sales,predict(lm.fit))

#Mean Absolute Percent Error

mape(Carseats$Sales,predict(lm.fit))

#Manualmente
sum(abs((Carseats$Sales-predict(lm.fit))/Carseats$Sales))/nrow(Carseats)

#Graficos de error

qqnorm(lm.fit$residuals)
qqline(lm.fit$residuals)

par(mfrow = c(2, 2))
plot(lm.fit)

#### Seleccion de mejores predictores ####
#R va a seleccionar las mejores exogenas

#Modelo original 
modelo <- lm(medv ~ ., Boston )

summary(modelo)

#Seleccion del mejor modelo
mejormodelo = step(object = modelo,
                   direction = "both", trace = 1) #both = metodo mixto

summary(mejormodelo)

#Modelo en forma de tabla para visualizar
stargazer(mejormodelo, title = "modelo estimado", type = "text")

#Intervalo de confianza de la regresion
confint(mejormodelo)

#Hipotesis de normalidad
shapiro.test(mejormodelo$residuals)

#Importancia de los regresores

datos_estandarizados = scale(Boston) #scale tipifica 
 
datos_estandarizados = as.data.frame(datos_estandarizados) #tabla de datos

eval(lm.fit$call[[2]]) #formula del modelo

lm.standard = lm(eval(lm.fit$call[[2]]),datos_estandarizados) 
#mismo modelo con datos tipificados

summary(lm.standard) #modelo tipificado

lm.standard$coefficients #parametros

which.max(abs(lm.standard$coefficients)) #variable mas influyente

#### Problemas del modelo lineal ####

lm.fit <- lm(medv~., data = Boston)

## No linealidad ##

par(mfrow = c(2, 2))
plot(lm.fit)

#Hay un patron en los residuos. No hay linealidad

#Solucionamos con una regresion polinomica

lm.fit5 <- lm(medv ~ poly(lstat , 3), data=Boston)

summary(lm.fit5)

par(mfrow = c(2, 2))
plot(lm.fit5)

comparacion = tibble(Boston, 'e.lineal'=predict(lm.fit),
                     'e.polinomica'=predict(lm.fit5))

#Grafico

ggplot(data = comparacion) +
  geom_point(mapping = aes(x = lstat, y = medv)) +
  geom_line(mapping = aes(x=lstat, y = e.lineal), colour = 'red',lwd=0.75) +
  geom_line(mapping = aes(x=lstat, y = e.polinomica), colour = 'blue',lwd=0.75) +
  ggtitle('Comparacion')+
  labs(x='X',y='Y')

## Autocorrelacion ##

#Contraste de Durbin Watson

#library(lmtest)
dwtest(lm.fit,alternative ="two.sided",iterations = 1000)
#Existe autocorrelacion, rechaza Ho: p = 0

#library(car)
durbinWatsonTest(lm.fit,simulate = TRUE,reps = 1000)
#Existe autocorrelacion, rechaza Ho: p = 0

## Heterocedasticidad ##

residuos = tibble(Boston,'estimaciones'=lm.fit$fitted.values,
                  'residuos'=lm.fit$residuals)

ggplot(data = residuos) +
  geom_point(mapping = aes(x = estimaciones, y = residuos)) +
  geom_smooth(mapping = aes(x = estimaciones, y = residuos), 
              colour = 'blue',lwd=0.5, se=FALSE)+
  ggtitle('Grafico de residuos')+
  labs(x='Fitted Values',y='Residuals')
  
#Goldfelt Quant test

gqtest(lm.fit, order.by = NULL, data = Boston, fraction = 10)
#Rechaza Ho: Homocedasticidad; hay heterocedasticidad

#Test de White

prueba_white<-bptest(lm.fit,data = Boston)
#Rechaza Ho: Homocedasticidad; hay heterocedasticidad

#Breusch-Pagan test

bptest(lm.fit, varformula = NULL, studentize = TRUE, data = Boston)
#Rechaza Ho: Homocedasticidad; hay heterocedasticidad

## Valores atipicos ## 

#Estadistico hat (h) y distancia de cook

#Leverages (hat): influyentes aquellas con hat mayor 2.5((p+1)/n)
#Distancia Cook (cook.d): Se consideran influyentes valores superiores a 1.

influencePlot(lm.fit)

(h = influencePlot(lm.fit)[2]) 

rango.hat(Boston)

which(h>rango.hat(Boston)) #Cuales superan el rango hat

h[which(h>rango.hat(Boston)),]

#Distancia de Mahalanobis

maha.F <- mahalanobis(Boston, 
                      center = colMeans(Boston),
                      cov = cov(Boston)) #colmeans es la media por columnas (variables)

plot(maha.F, pch=22, col="blue",main="Outliers")
text(x=1:length(Boston[,1]),y=maha.F, 1:length(Boston[,1]), pos=1, col="black")
abline(h=(mean(maha.F)+2*sd(maha.F)),col="red")

which(maha.F>(mean(maha.F)+2*sd(maha.F))) #Cuales superan media mas 2*sd

#Quitar los outlier
quitar<-which(maha.F>(mean(maha.F)+2*sd(maha.F)))
datos_sin_outlier = Boston[-quitar,]

#Distancia de Cook

cooksd <- cooks.distance(lm.fit)

plot(cooksd)
text(x=1:length(Boston[,1]),y=cooksd, 1:length(Boston[,1]), pos=1, col="blue")
abline(h=(mean(cooksd)+2*sd(cooksd)),col="red")

which.max(cooksd)
which(cooksd>(mean(cooksd)+2*sd(cooksd))) #Cuales superan 5 veces la media

## Multicolinealidad ## 

vif(lm.fit) #VIF

which.max(vif(lm.fit)) #Cual es el mayor

## Cambio Estructural ##

#Chow breakpoint test

sctest(lm.fit) #Ho no hay cambio estructural

#Busqueda del momento de cambio

BP<-breakpoints(medv ~ 1,data=Boston) #Momento cambio estructural

BP #Optimal n-segment partition se debe partir la muestra en n segmentos

summary(BP)

plot(BP)

#F breakpoint statistic 
#Para todo punto se calcula el test de Chow
#Ho: Bo = B1, mas F mas cerca de rechazar

fs <- Fstats(medv ~ 1,data=Boston)

plot(fs)

breakpoints(fs)

lines(breakpoints(fs))

#Podemos ver las lineas de cambio estructural:

ggplot(data = Boston) +
  geom_smooth(mapping = aes(y = medv, x = c(1:length(medv)))) +
  geom_vline(xintercept = breakpoints(fs)$breakpoints)+
  ggtitle('Cambios de estructura')+
  labs(x='t',y='Y')

#Chow Breakpoint Test

sctest(lm.fit, type = "Chow", point = 373) #Ho: Bo = B1

#Grafico regresion lineal con intervalos (R base)

reg.conf.intervals <- function(x, y) {
  n <- length(y) # Find length of y to use as sample size
  lm.model <- lm(y ~ x) # Fit linear model
  
  # Extract fitted coefficients from model object
  b0 <- lm.model$coefficients[1]
  b1 <- lm.model$coefficients[2]
  
  # Find SSE and MSE
  sse <- sum((y - lm.model$fitted.values)^2)
  mse <- sse / (n - 2)
  
  t.val <- qt(0.975, n - 2) # Calculate critical t-value
  
  # Fit linear model with extracted coefficients
  x_new <- 1:max(x)
  y.fit <- b1 * x_new + b0
  
  # Find the standard error of the regression line
  se <- sqrt(sum((y - y.fit)^2) / (n - 2)) * sqrt(1 / n + (x - mean(x))^2 / sum((x - mean(x))^2))
  
  # Fit a new linear model that extends past the given data points (for plotting)
  x_new2 <- 1:max(x + 100)
  y.fit2 <- b1 * x_new2 + b0
  
  # Warnings of mismatched lengths are suppressed
  slope.upper <- suppressWarnings(y.fit2 + t.val * se)
  slope.lower <- suppressWarnings(y.fit2 - t.val * se)
  
  # Collect the computed confidence bands into a data.frame and name the colums
  bands <- data.frame(cbind(slope.lower, slope.upper))
  colnames(bands) <- c('Lower Confidence Band', 'Upper Confidence Band')
  
  # Plot the fitted linear regression line and the computed confidence bands
  plot(x, y, cex = 1.75, pch = 21, bg = 'gray')
  lines(y.fit2, col = 'black', lwd = 2)
  lines(bands[1], col = 'blue', lty = 2, lwd = 2)
  lines(bands[2], col = 'blue', lty = 2, lwd = 2)
  
  return(bands)
}

conf.intervals <- reg.conf.intervals(cars$speed, cars$dist)

#CUMSUM

#Calcular CUMSUM

cumsum = efp(lm.fit, data=Boston, type="OLS-CUSUM")

summary(cumsum)

#Fronteras

bound.cumsum <- boundary(cumsum, alpha=0.05)

plot(cumsum)

#Violacion de las fronteras

which(cumsum$process>bound.cumsum)

#Test
sctest(cumsum)

#Transformacion logaritmica

lm.fit.log = lm(mpg ~ ., log(Auto[,-9])) #Aplicamos logaritmos

summary(lm.fit.log)

#Que variables son significativas
which(summary(lm.fit.log)$coefficients[, "Pr(>|t|)"]<0.025)

tStest(summary(lm.fit.log)$coefficients[,4])

plot(lm.fit.log) #No hay heterocedasticidad ni patron en el error

#Todas las posibles interacciones de segundo grado
lm.fit4 = lm(mpg ~ .^2, Auto[,-9])

summary(lm.fit4)

(coefit4 = summary(lm.fit4)$coefficients)

which(coefit4[, "Pr(>|t|)"]<0.025) #Cuales variables son significativas a un 5%

tStest(summary(lm.fit4)$coefficients[,4])

plot(lm.fit4)


